ssl/ssl_x509.cc

```diff
@@ -981,6 +981,134 @@ int SSL_add1_chain_cert(SSL *ssl, X509 *x509) {
   return ssl_cert_add1_chain_cert(ssl->config->cert.get(), x509);
 }
 
+#define is_flag_set(flags, query) (flags & query)
+
+static int ssl_build_cert_chain(CERT *cert, X509_STORE *cert_store, int flags) {
+  assert(cert_store);
+  if (!ssl_cert_check_cert_private_keys_usage(cert)) {
+    return 0;
+  }
+
+  CERT_PKEY *cert_pkey = &cert->cert_private_keys[cert->cert_private_key_idx];
+  CRYPTO_BUFFER *leaf_buffer =
+      sk_CRYPTO_BUFFER_value(cert_pkey->chain.get(), 0);
+  if (leaf_buffer == nullptr) {
+    OPENSSL_PUT_ERROR(SSL, SSL_R_NO_CERTIFICATE_SET);
+    return 0;
+  }
+  UniquePtr<X509> leaf(X509_parse_from_buffer(leaf_buffer));
+  if (leaf == nullptr) {
+    OPENSSL_PUT_ERROR(SSL, ERR_R_X509_LIB);
+    return 0;
+  }
+
+  UniquePtr<X509_STORE> store(X509_STORE_new());
+  UniquePtr<X509_STORE_CTX> store_ctx(X509_STORE_CTX_new());
+  if (store == nullptr || store_ctx == nullptr) {
+    OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  // Rearranging and check the chain: add everything to a store
+  UniquePtr<STACK_OF(X509)> untrusted(sk_X509_new_null());
+  if (is_flag_set(flags, SSL_BUILD_CHAIN_FLAG_CHECK)) {
+    // Push certs onto |X509_STORE|.
+    for (size_t i = 1; i < sk_CRYPTO_BUFFER_num(cert_pkey->chain.get()); i++) {
+      CRYPTO_BUFFER *buffer = sk_CRYPTO_BUFFER_value(cert_pkey->chain.get(), i);
+      UniquePtr<X509> x509(X509_parse_from_buffer(buffer));
+      if (!x509 || !X509_STORE_add_cert(store.get(), x509.get())) {
+        return 0;
+      }
+    }
+    // Add end-entity certificate too: it might be self-signed.
+    if (!X509_STORE_add_cert(store.get(), leaf.get())) {
+      return 0;
+    }
+  } else {
+    // Use associated |cert_store| from |SSL_CTX|. Reference count added to
+    // avoid double freeing of |X509_STORE|.
+    store.reset(cert_store);
+    X509_STORE_up_ref(cert_store);
+
+    if (is_flag_set(flags, SSL_BUILD_CHAIN_FLAG_UNTRUSTED)) {
+      // Push certs onto untrusted stack.
+      for (size_t i = 1; i < sk_CRYPTO_BUFFER_num(cert_pkey->chain.get());
+           i++) {
+        CRYPTO_BUFFER *buffer =
+            sk_CRYPTO_BUFFER_value(cert_pkey->chain.get(), i);
+        UniquePtr<X509> x509(X509_parse_from_buffer(buffer));
+        if (!x509 || !PushToStack(untrusted.get(), std::move(x509))) {
+          return 0;
+        }
+      }
+    }
+  }
+
+  if (!X509_STORE_CTX_init(store_ctx.get(), store.get(), leaf.get(),
+                           untrusted.get())) {
+    OPENSSL_PUT_ERROR(SSL, ERR_R_X509_LIB);
+    return 0;
+  }
+
+  bool ignore_error = false;
+  if (X509_verify_cert(store_ctx.get()) <= 0) {
+    // Fail if |SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR| is not set.
+    if(!is_flag_set(flags, SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR)) {
+      OPENSSL_PUT_ERROR(SSL, SSL_R_CERTIFICATE_VERIFY_FAILED);
+      ERR_add_error_data(2, "Verify error:",
```

Comment:
what's the significance of 2 here?

Reply:
The first argument of `ERR_add_error_data` indicates how many subsequent arguments (error strings) there are to be passed in.
https://github.com/aws/aws-lc/blob/05449f5e8132496360df1c799535a2dd633f8d5e/include/openssl/err.h#L458-L461